### PR TITLE
get contract sources from contract definition

### PR DESCRIPTION
**Pull request by technophile-04**
## Description

Fixes : 

https://github.com/scaffold-eth/scaffold-eth-2/pull/564#issuecomment-1781833141

https://github.com/scaffold-eth/scaffold-eth-2/pull/564#issuecomment-1763430600

### CODE REVIEW

**Comment by technophile-04** [comment]
Summary of the solution : 

If you look at `sources` which is generated by `hardhat-deploy` for contract `E` are : 

<details>
<summary>
 Contract `E` sources : 
</summary>

```js
{
  'contracts/A.sol': {
    content: '//SPDX-License-Identifier: MIT\n' +
      'pragma solidity >=0.8.0 <0.9.0;\n' +
      '\n' +
      'contract A {\n' +
      '\tfunction foo() public pure virtual returns (string memory) {\n' +
      '\t\treturn "A";\n' +
      '\t}\n' +
      '}\n',
    keccak256: '0xf1fd7d3a46e81bfb4bb4c45788eb1f09c8991ebde86ce57716c2da7725713c69',
    license: 'MIT'
  },
  'contracts/B.sol': {
    content: '//SPDX-License-Identifier: MIT\n' +
      'pragma solidity >=0.8.0 <0.9.0;\n' +
      '\n' +
      'import "./A.sol";\n' +
      '\n' +
      "// Contracts inherit other contracts by using the keyword 'is'.\n" +
      'contract B is A {\n' +
      '\t// Override A.foo()\n' +
      '\tfunction foo() public pure virtual override returns (string memory) {\n' +
      '\t\treturn "B";\n' +
      '\t}\n' +
      '}\n',
    keccak256: '0x694df20286795fd4f09730c2157578d071f13b57fdca7940a4e4111b1c677583',
    license: 'MIT'
  },
  'contracts/C.sol': {
    content: '//SPDX-License-Identifier: MIT\n' +
      'pragma solidity >=0.8.0 <0.9.0;\n' +
      '\n' +
      'import "./A.sol";\n' +
      '\n' +
      'contract C is A {\n' +
      '\t// Override A.foo()\n' +
      '\tfunction foo() public pure virtual override returns (string memory) {\n' +
      '\t\treturn "C";\n' +
      '\t}\n' +
      '}\n',
    keccak256: '0x55497f031a5fee0c152747fa25bf6c8c79551ad61b455b209e652c1588126dbc',
    license: 'MIT'
  },
  'contracts/E.sol': {
    content: '//SPDX-License-Identifier: MIT\n' +
      'pragma solidity >=0.8.0 <0.9.0;\n' +
      '\n' +
      'import "./B.sol";\n' +
      'import "./C.sol";\n' +
      '\n' +
      'contract E is C, B {\n' +
      '\t// E.foo() returns "B"\n' +
      '\t// since B is the right most parent contract with function foo()\n' +
      '\tfunction foo() public pure override(C, B) returns (string memory) {\n' +
      '\t\treturn super.foo();\n' +
      '\t}\n' +
      '}\n',
    keccak256: '0x777c64b583bb7ae1eca9463b379a72811bbb31be28a590c61154a05395a4971f',
    license: 'MIT'
  }
}

```

</details>

In the previous solution we were just mapping over the above object keys using `object.keys(sources)` and as we can see in the above object `contract/C.sol` comes after `contract/B.sol` therefore `contract/C.sol` was overriding it. 

In a new solution instead of relying on object keys of `sources` we instead derive `actualSources` from this : 
`"contract E is C, B {\n"` that are `[C, B]` and hence always right most parent will override it 

This also nicely solves https://github.com/scaffold-eth/scaffold-eth-2/pull/564#issuecomment-1763430600 since now we are getting "actual real sources" from contract definition directly. 

Lol not that good explanation but if anyone wants to play around you can check Rinat's branch https://github.com/scaffold-eth/scaffold-eth-2/tree/inherited-edge-case (Here I have put helpful console.log to make sense) 🙌

File: packages/hardhat/deploy/99_generateTsAbis.ts [hunk 1]
```diff
@@ -27,20 +27,43 @@ function getContractNames(path: string) {
...
 
-function getInheritedFunctions(sources: Record<string, any>, contractName: string) {
-  const inheritedFunctions = {} as Record<string, any>;
+function getAcutalSourcesForContract(sources: Record<string, any>, contractName: string) {
   for (const sourcePath of Object.keys(sources)) {
     const sourceName = sourcePath.split("/").pop()?.split(".sol")[0];
     if (sourceName === contractName) {
-      continue;
+      const contractContent = sources[sourcePath].content as string;
+      const regex = /contract\s+(\w+)\s+is\s+([^{}]+)\{/;
```
Comment (technophile-04) [inline packages/hardhat/deploy/99_generateTsAbis.ts:35]: Also tested this regex nicely with different cases eg ;

`contract A is      B,C{`

`contract A is B,C{`

`contract A is B,C      {` 

`contract A     is  B,                       C {` 

etc just to be sure the spaces don't mess it up and it nicely returns  [B,C] every time 🙌

File: packages/hardhat/deploy/99_generateTsAbis.ts [hunk 2]
```diff
@@ -27,20 +27,43 @@ function getContractNames(path: string) {
     .map(dirent => dirent.name.split(".")[0]);
 }
 
-function getInheritedFunctions(sources: Record<string, any>, contractName: string) {
-  const inheritedFunctions = {} as Record<string, any>;
+function getAcutalSourcesForContract(sources: Record<string, any>, contractName: string) {
```
Comment (rin-st) [inline packages/hardhat/deploy/99_generateTsAbis.ts]: Actual

File: packages/hardhat/deploy/99_generateTsAbis.ts [hunk 3]
```diff
@@ -27,20 +27,43 @@ function getContractNames(path: string) {
...
+      const regex = /contract\s+(\w+)\s+is\s+([^{}]+)\{/;
+      const match = contractContent.match(regex);
+
+      if (match) {
+        const inheritancePart = match[2];
+        // Split the inherited contracts by commas to get the list of inherited contracts
+        const inheritedContracts = inheritancePart.split(",").map(contract => `${contract.trim()}.sol`);
+
+        return inheritedContracts;
+      } else return [];
```
Comment (rin-st) [inline packages/hardhat/deploy/99_generateTsAbis.ts]: you can skip `else`, just `return []`

**Review by rin-st — APPROVED** [review]

**Comment by rin-st** [comment]
Nice solution! Tested it and it works!

Left two small commments. Approve, thanks!

**Comment by FilipHarald** [comment]
Thank you for your help @technophile-04 @rin-st . I had a look and it looks great. :+1:

**Pull request merged**